### PR TITLE
feat(ethexe): zero program state; handle exec balances topups; handle unrecognised mirrors; refactor

### PR DESCRIPTION
**Pull request by breathx**

### CODE REVIEW

File: ethexe/runtime/common/src/state.rs [hunk 1]
```diff
@@ -111,6 +121,41 @@ pub struct ProgramState {
     pub executable_balance: Value,
 }
 
+impl ProgramState {
+    pub const fn zero() -> Self {
+        Self {
```
Comment (StackOverflowExcept1on) [inline ethexe/runtime/common/src/state.rs:126]: mb use derive default?

File: ethexe/processor/src/lib.rs [hunk 2]
```diff
@@ -121,8 +109,7 @@ impl Processor {
         Ok(true)
     }
 
-    // TODO: deal with params on smart contract side.
-    pub fn handle_new_program(&mut self, program_id: ProgramId, code_id: CodeId) -> Result<H256> {
+    pub fn handle_new_program(&mut self, program_id: ProgramId, code_id: CodeId) -> Result<()> {
```
Comment (StackOverflowExcept1on) [inline ethexe/processor/src/lib.rs:112]: mb check here that the program does not exist in db?

File: ethexe/processor/src/lib.rs [hunk 3]
```diff
@@ -272,11 +218,9 @@ impl Processor {
     ) -> Result<Vec<LocalOutcome>> {
         log::debug!("Processing upload code {code_id:?}");
 
-        if code_id != CodeId::generate(code) || self.handle_new_code(code)?.is_none() {
-            Ok(vec![LocalOutcome::CodeRejected(code_id)])
-        } else {
-            Ok(vec![LocalOutcome::CodeApproved(code_id)])
-        }
+        let valid = !(code_id != CodeId::generate(code) || self.handle_new_code(code)?.is_none());
```
Comment (StackOverflowExcept1on) [inline ethexe/processor/src/lib.rs:221]:
```suggestion
        let valid = code_id == CodeId::generate(code) && self.handle_new_code(code)?.is_some();

```

**Review by StackOverflowExcept1on — APPROVED** [review]

File: ethexe/processor/src/lib.rs (hunk 3 above)
Comment (StackOverflowExcept1on) [reply ethexe/processor/src/lib.rs:221]: what if `self.handle_new_code(code)?` returns an error? user won't get `LocalOutcome::CodeValidated`

File: ethexe/runtime/common/src/state.rs (hunk 1 above)
Comment (breathx) [reply ethexe/runtime/common/src/state.rs:126]: I think const fns with more intuitive about usage naming for the case looks better

File: ethexe/processor/src/lib.rs (hunk 2 above)
Comment (breathx) [reply ethexe/processor/src/lib.rs:112]: program -> code_id matching does exactly the thing

**Pull request merged**